### ISSUE TITLE
Best First Search priority queue presents a weird behavior

### ISSUE THREAD

**Issue by kecol** [aimacode/aima-python#919]
Looks like this modification to PriorityQueue.__contains__() method makes the Best First Search original implementation fails with the frontier management. The queue will not work as expected by BFS algorithm and the frontier will grow in size every time a repeated situation/state is found.

In class PriotityQueue
https://github.com/aimacode/aima-python/blob/5889656bba37e8fb7b1464bb1b9c8e0a9a9919f8/utils.py#L726-L728

Inside best_first_graph_search method (lines 276 and 278). 
https://github.com/aimacode/aima-python/blob/5889656bba37e8fb7b1464bb1b9c8e0a9a9919f8/search.py#L275-L282

**Comment by rajatjain1997** [comment]
You are correct, frontier size will increase not only in the case of BFS but also in the instances of A* algorithm where the heuristic function has different values for the same node depending on the path taken. Thanks for pointing this out. I will try to resolve this issue ASAP :-).

**Comment by dmeoli** [comment]
Resolved by #1196 (commit `b57c2e5`). The weird behavior was caused by `PriorityQueue.__contains__` comparing the whole `(f, item)` tuple, so a same-state node with a different f-value failed the membership test and got appended as a duplicate (unbounded frontier growth) instead of replacing the costlier entry. `__contains__`/`__getitem__`/`__delitem__` now match on the item only (via `Node.__eq__`, state-based), so `best_first_graph_search`'s replacement works correctly. Verified on current master. Closing as resolved.
